File: packages/vmcp-dsl/src/patterns/filter.ts
/**
 * Filter pattern builder
 */

import type {
  PatternSpec,
  FilterSpec,
  FieldPredicate,
  PredicateOp,
  PredicateValue,
} from '../types.js';
import { path } from '../path-builder.js';

/**
 * Builder for filter patterns
 */
export class FilterBuilder {
  private predicate: Partial<FieldPredicate> = {};

  /**
   * Set the field to filter on
   */
  field(pathExpr: string): this {
    this.predicate.field = path(pathExpr);
    return this;
  }

  /**
   * Equal to
   */
  eq(value: string | number | boolean): this {
    this.predicate.op = 'eq';
    this.predicate.value = toPredicateValue(value);
    return this;
  }

  /**
   * Not equal to
   */
  ne(value: string | number | boolean): this {
    this.predicate.op = 'ne';
    this.predicate.value = toPredicateValue(value);
    return this;
  }

  /**
   * Greater than
   */
  gt(value: number): this {
    this.predicate.op = 'gt';
    this.predicate.value = { numberValue: value };
    return this;
  }

  /**
   * Greater than or equal
   */
  gte(value: number): this {
    this.predicate.op = 'gte';
    this.predicate.value = { numberValue: value };
    return this;
  }

  /**
   * Less than
   */
  lt(value: number): this {
    this.predicate.op = 'lt';
    this.predicate.value = { numberValue: value };
    return this;
  }

  /**
   * Less than or equal
   */
  lte(value: number): this {
    this.predicate.op = 'lte';
    this.predicate.value = { numberValue: value };
    return this;
  }

  /**
   * String contains
   */
  contains(value: string): this {
    this.predicate.op = 'contains';
    this.predicate.value = { stringValue: value };
    return this;
  }

  /**
   * Value in list
   */
  in(values: (string | number | boolean)[]): this {
    this.predicate.op = 'in';
    this.predicate.value = { listValue: values.map(toPredicateValue) };
    return this;
  }

  /**
   * Is null
   */
  isNull(): this {
    this.predicate.op = 'eq';
    this.predicate.value = { nullValue: true };
    return this;
  }

  /**
   * Is not null
   */
  isNotNull(): this {
    this.predicate.op = 'ne';
    this.predicate.value = { nullValue: true };
    return this;
  }

  /**
   * Build the filter pattern spec
   */
  build(): PatternSpec {
    if (!this.predicate.field) {
      throw new Error('Field is required');
    }
    if (!this.predicate.op) {
      throw new Error('Operator is required');
    }
    if (!this.predicate.value) {
      throw new Error('Value is required');
    }
    return { filter: { predicate: this.predicate as FieldPredicate } };
  }

  /**
   * Get the raw spec
   */
  spec(): FilterSpec {
    if (!this.predicate.field || !this.predicate.op || !this.predicate.value) {
      throw new Error('Incomplete filter predicate');
    }
    return { predicate: this.predicate as FieldPredicate };
  }
}

/**
 * Convert a primitive value to PredicateValue
 */
function toPredicateValue(value: string | number | boolean): PredicateValue {
  if (typeof value === 'string') {
    return { stringValue: value };
  }
  if (typeof value === 'number') {
    return { numberValue: value };
  }
  if (typeof value === 'boolean') {
    return { boolValue: value };
  }
  throw new Error(`Unsupported value type: ${typeof value}`);
}

/**
 * Create a filter pattern
 *
 * @example
 * const highScoreFilter = filter()
 *   .field('$.score')
 *   .gt(0.7)
 *   .build();
 *
 * const typeFilter = filter()
 *   .field('$.type')
 *   .in(['pdf', 'html', 'doc'])
 *   .build();
 */
export function filter(): FilterBuilder {
  return new FilterBuilder();
}

/**
 * Shorthand for creating a simple filter
 *
 * @example
 * const spec = filterBy('$.score', 'gt', 0.5);
 */
export function filterBy(
  field: string,
  op: PredicateOp,
  value: string | number | boolean | (string | number | boolean)[]
): PatternSpec {
  const builder = new FilterBuilder().field(field);

  switch (op) {
    case 'eq':
      return builder.eq(value as string | number | boolean).build();
    case 'ne':
      return builder.ne(value as string | number | boolean).build();
    case 'gt':
      return builder.gt(value as number).build();
    case 'gte':
      return builder.gte(value as number).build();
    case 'lt':
      return builder.lt(value as number).build();
    case 'lte':
      return builder.lte(value as number).build();
    case 'contains':
      return builder.contains(value as string).build();
    case 'in':
      return builder.in(value as (string | number | boolean)[]).build();
    default:
      throw new Error(`Unknown operator: ${op}`);
  }
}

File: crates/agentgateway/src/mcp/registry/patterns/filter.rs
// Filter pattern types

use serde::{Deserialize, Serialize};

/// FilterSpec filters array elements based on a predicate
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FilterSpec {
	/// The predicate to evaluate for each element
	pub predicate: FieldPredicate,
}

/// A predicate that compares a field value
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FieldPredicate {
	/// JSONPath to the field to evaluate
	pub field: String,

	/// Comparison operator
	pub op: String,

	/// Value to compare against
	pub value: PredicateValue,
}

impl FieldPredicate {
	/// Create a new predicate
	pub fn new(field: impl Into<String>, op: impl Into<String>, value: PredicateValue) -> Self {
		Self { field: field.into(), op: op.into(), value }
	}

	/// Create an equality predicate
	pub fn eq(field: impl Into<String>, value: impl Into<PredicateValue>) -> Self {
		Self::new(field, "eq", value.into())
	}

	/// Create a greater-than predicate
	pub fn gt(field: impl Into<String>, value: f64) -> Self {
		Self::new(field, "gt", PredicateValue::Number(value))
	}

	/// Create a less-than predicate
	pub fn lt(field: impl Into<String>, value: f64) -> Self {
		Self::new(field, "lt", PredicateValue::Number(value))
	}

	/// Create a contains predicate (for strings)
	pub fn contains(field: impl Into<String>, value: impl Into<String>) -> Self {
		Self::new(field, "contains", PredicateValue::String(value.into()))
	}
}

/// A value used in predicate comparisons
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum PredicateValue {
	/// String value
	StringValue(String),

	/// Numeric value
	NumberValue(f64),

	/// Boolean value
	BoolValue(bool),

	/// Null value
	NullValue(bool),

	/// List of values (for "in" operator)
	ListValue(Vec<PredicateValue>),
}

impl PredicateValue {
	/// Create a string value
	pub fn String(s: impl Into<String>) -> Self {
		PredicateValue::StringValue(s.into())
	}

	/// Create a number value
	pub fn Number(n: f64) -> Self {
		PredicateValue::NumberValue(n)
	}

	/// Create a boolean value
	pub fn Bool(b: bool) -> Self {
		PredicateValue::BoolValue(b)
	}

	/// Create a null value
	pub fn Null() -> Self {
		PredicateValue::NullValue(true)
	}

	/// Create a list value
	pub fn List(values: Vec<PredicateValue>) -> Self {
		PredicateValue::ListValue(values)
	}

	/// Convert to serde_json::Value for comparison
	pub fn to_json_value(&self) -> serde_json::Value {
		match self {
			PredicateValue::StringValue(s) => serde_json::Value::String(s.clone()),
			PredicateValue::NumberValue(n) => serde_json::json!(n),
			PredicateValue::BoolValue(b) => serde_json::Value::Bool(*b),
			PredicateValue::NullValue(_) => serde_json::Value::Null,
			PredicateValue::ListValue(values) => {
				serde_json::Value::Array(values.iter().map(|v| v.to_json_value()).collect())
			},
		}
	}
}

impl From<String> for PredicateValue {
	fn from(s: String) -> Self {
		PredicateValue::StringValue(s)
	}
}

impl From<&str> for PredicateValue {
	fn from(s: &str) -> Self {
		PredicateValue::StringValue(s.to_string())
	}
}

impl From<f64> for PredicateValue {
	fn from(n: f64) -> Self {
		PredicateValue::NumberValue(n)
	}
}

impl From<i64> for PredicateValue {
	fn from(n: i64) -> Self {
		PredicateValue::NumberValue(n as f64)
	}
}

impl From<bool> for PredicateValue {
	fn from(b: bool) -> Self {
		PredicateValue::BoolValue(b)
	}
}

#[cfg(test)]
mod tests {
	use super::*;

	#[test]
	fn test_parse_filter() {
		let json = r#"{
			"predicate": {
				"field": "$.score",
				"op": "gt",
				"value": { "numberValue": 0.7 }
			}
		}"#;

		let filter: FilterSpec = serde_json::from_str(json).unwrap();
		assert_eq!(filter.predicate.field, "$.score");
		assert_eq!(filter.predicate.op, "gt");
		assert!(matches!(filter.predicate.value, PredicateValue::NumberValue(n) if (n - 0.7).abs() < f64::EPSILON));
	}

	#[test]
	fn test_parse_predicate_string_value() {
		let json = r#"{
			"field": "$.source",
			"op": "eq",
			"value": { "stringValue": "arxiv" }
		}"#;

		let pred: FieldPredicate = serde_json::from_str(json).unwrap();
		assert!(matches!(pred.value, PredicateValue::StringValue(ref s) if s == "arxiv"));
	}

	#[test]
	fn test_parse_predicate_list_value() {
		let json = r#"{
			"field": "$.type",
			"op": "in",
			"value": {
				"listValue": [
					{ "stringValue": "pdf" },
					{ "stringValue": "html" }
				]
			}
		}"#;

		let pred: FieldPredicate = serde_json::from_str(json).unwrap();
		assert!(matches!(pred.value, PredicateValue::ListValue(_)));
	}

	#[test]
	fn test_predicate_builders() {
		let eq = FieldPredicate::eq("$.name", "test");
		assert_eq!(eq.op, "eq");

		let gt = FieldPredicate::gt("$.score", 0.5);
		assert_eq!(gt.op, "gt");

		let contains = FieldPredicate::contains("$.text", "keyword");
		assert_eq!(contains.op, "contains");
	}

	#[test]
	fn test_predicate_value_to_json() {
		assert_eq!(PredicateValue::String("hello").to_json_value(), serde_json::json!("hello"));
		assert_eq!(PredicateValue::Number(42.0).to_json_value(), serde_json::json!(42.0));
		assert_eq!(PredicateValue::Bool(true).to_json_value(), serde_json::json!(true));
		assert_eq!(PredicateValue::Null().to_json_value(), serde_json::Value::Null);
	}
}

